### ISSUE TITLE
deploy problem

### ISSUE THREAD

**Issue by ck** [weavejester/lein-beanstalk#1]
I'm playing around with your beanstalk plugin and are having some problems.

I created a small sample app (https://github.com/ck/hello-beanstalk) and tried to deploy it to Amazon Elastic Beanstalk. Unfortunately, I always get the following error:

```
$ lein beanstalk deploy :production
Generating war file
Created /Users/ck/code/active/hello-beanstalk/hb-1.0.0-SNAPSHOT-beanstalk.war
Uploading war file to S3
Feb 5, 2011 12:59:36 PM com.amazonaws.http.HttpClient execute
INFO: Sending Request: PUT https://lein-beanstalk.s3.amazonaws.com / Headers: (Authorization: AWS oiehuiwef:SMFEFBWEBEW=, Date: Sat, 05 Feb 2011 17:59:36 GMT, Content-Type: application/x-www-form-urlencoded; charset=utf-8, ) 
Feb 5, 2011 12:59:44 PM com.amazonaws.http.HttpClient handleErrorResponse
INFO: Received error response: Status Code: 409, AWS Request ID: 1312321312, AWS Error Code: BucketAlreadyExists, AWS Error Message: The requested bucket name is not available. The bucket namespace is shared by all users of the system. Please select a different name and try again., S3 Extended Request ID: sds
Exception in thread "main" Status Code: 409, AWS Request ID: 1312321312, AWS Error Code: BucketAlreadyExists, AWS Error Message: The requested bucket name is not available. The bucket namespace is shared by all users of the system. Please select a different name and try again., S3 Extended Request ID: sdadsa (NO_SOURCE_FILE:0)
    at clojure.lang.Compiler.eval(Compiler.java:5440)
    at clojure.lang.Compiler.eval(Compiler.java:5391)
    at clojure.core$eval.invoke(core.clj:2382)
    at clojure.main$eval_opt.invoke(main.clj:235)
    at clojure.main$initialize.invoke(main.clj:254)
    at clojure.main$script_opt.invoke(main.clj:270)
    at clojure.main$main.doInvoke(main.clj:354)
    at clojure.lang.RestFn.invoke(RestFn.java:513)
    at clojure.lang.Var.invoke(Var.java:385)
    at clojure.lang.AFn.applyToHelper(AFn.java:187)
    at clojure.lang.Var.applyTo(Var.java:482)
    at clojure.main.main(main.java:37)
Caused by: Status Code: 409, AWS Request ID: 26273D3A0130396E, AWS Error Code: BucketAlreadyExists, AWS Error Message: The requested bucket name is not available. The bucket namespace is shared by all users of the system. Please select a different name and try again., S3 Extended Request ID: dsdsdsa+dwMzZWqqY
    at com.amazonaws.http.HttpClient.handleErrorResponse(HttpClient.java:519)
    at com.amazonaws.http.HttpClient.execute(HttpClient.java:215)
    at com.amazonaws.services.s3.AmazonS3Client.createBucket(AmazonS3Client.java:545)
    at com.amazonaws.services.s3.AmazonS3Client.createBucket(AmazonS3Client.java:492)
    at leiningen.beanstalk$upload_war.invoke(beanstalk.clj:41)
    at leiningen.beanstalk$deploy.invoke(beanstalk.clj:69)
    at clojure.lang.AFn.applyToHelper(AFn.java:165)
    at clojure.lang.AFn.applyTo(AFn.java:151)
    at clojure.core$apply.invoke(core.clj:542)
    at leiningen.beanstalk$beanstalk.doInvoke(beanstalk.clj:81)
    at clojure.lang.RestFn.invoke(RestFn.java:443)
    at clojure.lang.Var.invoke(Var.java:373)
    at clojure.lang.AFn.applyToHelper(AFn.java:169)
    at clojure.lang.Var.applyTo(Var.java:482)
    at clojure.core$apply.invoke(core.clj:542)
    at leiningen.core$apply_task.invoke(core.clj:191)
    at leiningen.core$_main.doInvoke(core.clj:250)
    at clojure.lang.RestFn.applyTo(RestFn.java:138)
    at clojure.core$apply.invoke(core.clj:542)
    at leiningen.core$_main.invoke(core.clj:255)
    at user$eval175.invoke(NO_SOURCE_FILE:1)
    at clojure.lang.Compiler.eval(Compiler.java:5424)
    ... 11 more
```

I checked before and after the deploy and there is no bucket with that name.

Any pointers are appreciated,

-ck

**Comment by weavejester** [comment]
Well, lein-beanstalk is a work in progress. I've only recently begun development on it, no versions have been released, and it only partially works under some very specific circumstances.

I'd appreciate any help in developing it, but otherwise it'll probably be a week or two before it's complete enough to be usable.

**Comment by ck** [comment]
Fair enough. 

I more then happy to help with code, docs, and/or example. Please let me know if there is anything particular I can take a look at.

-ck

**Comment by weavejester** [comment]
Well, when I run lein-beanstalk, the "create" parts work the first time, but the "update" parts aren't really there. The idea is that `lein beanstalk deploy` should create environments and application versions if they don't already exist, or update the existing ones if they do.

I'm not sure why it doesn't work at _all_ for you, but I have some guesses. First, the way I define the :aws configuration is as follows:

```
:aws {:access-key "XXXXXXXXXXXXXXXXX"
        :secret-key "XXXXXXXXXXXXXXX"
        :beanstalk {:environments [hello-jack-dev]}})
```

So I'm using symbols, rather than keywords. Maybe that's got something to do with it? Perhaps the ":" of the keyword is leaking through into the URL.

Another thing that might be causing it is that the environment name currently needs to be unique, since by default Amazon uses it for the subdomain. I'm not entirely sure how this works, since it seems to make it unique if it's not alread by adding a random string onto the end of the subdomain. But maybe it's objecting to "production"?

And when I run the deploy command, I use:

```
lein beanstalk deploy hello-jack-dev
```

Try all that and see if you can get it deploying at least once. If you can, then there's just the updating to work on (and the docs, error messages, etc.!)

**Comment by ck** [comment]
Playing around with it more I found the root cause for my first problem (and now feel stupid for not read the clear error message 'The bucket namespace is shared by all users of the system'):

The bucket name itself is hard-coded to 'lein-beanstalk'. In S3 the bucket names are 'global' and assuming you have a bucket with your name under your account, I therefore can not create one with the same.

I'll try again with the symbol instead of the keyword and take a look at update as well.

If you don't mind, I will continue playing around with it in my fork and send you a message for review. Please keep in mind this will be by first plugin. I'll try my best, but there probably will be a lot 'bad' code.

-ck

**Comment by weavejester** [comment]
Ahh - I misunderstood the API.

I guess the bucket-name should default to a unique but descriptive value like the project name, with perhaps a prefix, e.g. "lein-beanstalk/<project name>". It should probably also be an option under the :beanstalk key.

**Comment by ck** [comment]
James:

I have another stupid problem:
I tried to deploy my hello-beanstalk app manually to AWS Beanstalk, but without success. First, I got an error indicating that jar could not be loaded and the offending class is `javax/servlet/Servlet.class`. After I removed `servlet-api-2.5.jar` from the lib folder and rebuild the war, I did not get the error anymore, but the app still did not deploy properly (i.e. I still have a red square in the UI). 

Did you ever try a deploy via the web site and ran into the same problem?

-ck

**Comment by weavejester** [comment]
Did you generate the war file to be uploaded using "lein ring uberwar"?

**Comment by ck** [comment]
I tried using the war created by
1. `lein ring uberwar`
2. `lein beanstalk deploy <env>`

I had trouble with both (I just double checked that 1 still does not work). The strange part is that the [log](https://gist.github.com/813405) file looks clean. On the other hand,I am able to deploy the same war to a local Tomcat6 instance by simply copying it to the webapps folder. 

Btw, are you on any IRC channel? Might be easier then using the git issue system.

-ck

**Comment by weavejester** [comment]
IRC might be a good idea. I'll try and be on the #clojure freenode channel.

Regarding the issue, there are two separate problems. The `servlet-api-2.5.jar` error occurs because there is a conflict between the servlet-api jar in the war file, and the one used by Tomcat. This is just a warning message, and does not prevent the web application from running. However, in the interests of not generating misleading warnings, I've updated lein-ring to specifically exclude any servlet-api jars (version 0.3.2).

The next problem is the red square in the UI. After some investigation, I found that Elastic Beanstalk checks to see if your web application is running by sending a HTTP HEAD request to "/". It expects a 200; if it doesn't get one, it assumes the web application failed to deploy. So this means that you need to have the following Compojure route:

```
(HEAD "/" [] "")
```

Then Elastic Beanstalk will register your deployment as having succeeded. I suspect that I'll have to think about making GETs add an implicit HEAD. According to the HTTP RFC, the response from a HEAD and a GET request should be identical, except the HEAD lacks a body.

**Comment by ck** [comment]
You are right. I tried your `compojure-example` application and even though it was marked red it still worked when deployed manually. Adding the HEAD route got rid of the red.

Additionally, I pulled `compojure-example` again containing the latest lein-ring, compojure, and hiccup version, and can confirm that the `servlet-api-2.5.jar` problem is resolved.   

With that done, I'll get back looking into lein-beanstalk.

Btw, thanks for all your help. I appreciate it.

-ck
